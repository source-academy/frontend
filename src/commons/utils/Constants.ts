import { Chapter, Variant } from 'js-slang/dist/types';
import moment, { Moment } from 'moment';

function isTrue(value?: string, defaultTo?: boolean): boolean {
  return typeof value === 'undefined' && typeof defaultTo !== 'undefined'
    ? defaultTo
    : typeof value === 'string' && value.toUpperCase() === 'TRUE';
}

const isTest = process.env.NODE_ENV === 'test';

const sourceAcademyVersion = process.env.REACT_APP_VERSION || 'local';
const sourceAcademyEnvironment = process.env.REACT_APP_ENVIRONMENT || 'dev';
const sourceAcademyDeploymentName = process.env.REACT_APP_DEPLOYMENT_NAME || 'Source Academy';
const showResearchPrompt = isTest || isTrue(process.env.REACT_APP_SHOW_RESEARCH_PROMPT);
const backendUrl = process.env.REACT_APP_BACKEND_URL;
const cadetLoggerUrl = isTest ? undefined : process.env.REACT_APP_CADET_LOGGER;
const cadetLoggerInterval = parseInt(process.env.REACT_APP_CADET_LOGGER_INTERVAL || '10000', 10);
const useBackend = !isTest && isTrue(process.env.REACT_APP_USE_BACKEND);
const defaultSourceChapter = Chapter.SOURCE_4;
const defaultSourceVariant = Variant.DEFAULT;
const defaultQuestionId = 0;
const maxBrowseIndex = 50;
const mobileBreakpoint = 768;
const urlShortenerBase = process.env.REACT_APP_URL_SHORTENER_BASE;
const urlShortenerSignature = process.env.REACT_APP_URL_SHORTENER_SIGNATURE;
const moduleBackendUrl = process.env.REACT_APP_MODULE_BACKEND_URL || 'modules';
const sharedbBackendUrl = process.env.REACT_APP_SHAREDB_BACKEND_URL || '';
const playgroundOnly = !isTest && isTrue(process.env.REACT_APP_PLAYGROUND_ONLY, true);
const enableGitHubAssessments = isTest || isTrue(process.env.REACT_APP_ENABLE_GITHUB_ASSESSMENTS);
const sentryDsn = process.env.REACT_APP_SENTRY_DSN;
const googleClientId = process.env.REACT_APP_GOOGLE_CLIENT_ID;
const googleApiKey = process.env.REACT_APP_GOOGLE_API_KEY;
const googleAppId = process.env.REACT_APP_GOOGLE_APP_ID;
const githubClientId = process.env.REACT_APP_GITHUB_CLIENT_ID || '';
const githubOAuthProxyUrl = process.env.REACT_APP_GITHUB_OAUTH_PROXY_URL || '';
const sicpBackendUrl =
  process.env.REACT_APP_SICPJS_BACKEND_URL || 'https://sicp.sourceacademy.org/';
const workspaceSettingsLocalStorageKey = 'workspace-settings';

// For achievements feature (CA - Continual Assessment)
// TODO: remove dependency of the ca levels on the env file
const caFulfillmentLevel = isTest
  ? 24
  : parseInt(process.env.REACT_APP_CA_FULFILLMENT_LEVEL || '0');

const authProviders: Map<string, { name: string; endpoint: string; isDefault: boolean }> =
  new Map();

for (let i = 1; ; ++i) {
  const id = process.env[`REACT_APP_OAUTH2_PROVIDER${i}`];
  if (!id) {
    break;
  }

  const name = process.env[`REACT_APP_OAUTH2_PROVIDER${i}_NAME`] || 'Unnamed provider';
  const endpoint = process.env[`REACT_APP_OAUTH2_PROVIDER${i}_ENDPOINT`] || '';

  authProviders.set(id, { name, endpoint, isDefault: i === 1 });
}

const disablePeriods: Array<{ start: Moment; end: Moment; reason?: string }> = [];

if (!isTest) {
  for (let i = 1; ; ++i) {
    const startStr = process.env[`REACT_APP_DISABLE${i}_START`];
    const endStr = process.env[`REACT_APP_DISABLE${i}_END`];
    if (!startStr || !endStr) {
      break;
    }
    const reason = process.env[`REACT_APP_DISABLE${i}_REASON`];
    const start = moment(startStr);
    const end = moment(endStr);
    if (end.isBefore(start) || moment().isAfter(end)) {
      continue;
    }
    disablePeriods.push({ start, end, reason });
  }
}

export enum Links {
  githubIssues = 'https://github.com/source-academy/frontend/issues',
  githubOrg = 'https://github.com/source-academy',
  about = 'https://about.sourceacademy.org',

  moduleDetails = 'https://www.comp.nus.edu.sg/~cs1101s',
  luminus = 'https://luminus.nus.edu.sg/modules/41d42e9a-5880-43b5-8ee6-75f5a41355e3/announcements/active',
  piazza = 'https://piazza.com/class/kas136yscf8605',

  resourcesForEducators = 'https://about.sourceacademy.org/educator/README.html',
  resourcesForLearners = 'https://about.sourceacademy.org/learner/README.html',

  sourceAcademyAssets = 'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/',
  sourceDocs = 'https://docs.sourceacademy.org/',
  techSVC = 'mailto:techsvc@comp.nus.edu.sg',
  techSVCNumber = '6516 2736',
  textbook = 'https://sourceacademy.org/sicpjs/',
  playground = 'https://sourceacademy.org/playground',
  textbookChapter2_2 = 'https://sourceacademy.org/sicpjs/2.2',
  textbookChapter3_2 = 'https://sourceacademy.org/sicpjs/3.2',
  aceHotkeys = 'https://github.com/ajaxorg/ace/wiki/Default-Keyboard-Shortcuts',
  sourceHotkeys = 'https://github.com/source-academy/frontend/wiki/Source-Academy-Keyboard-Shortcuts',

  source_1 = 'https://docs.sourceacademy.org/source_1/',
  source_1_Lazy = 'https://docs.sourceacademy.org/source_1_lazy/',
  source_1_Wasm = 'https://docs.sourceacademy.org/source_1_wasm/',
  source_2 = 'https://docs.sourceacademy.org/source_2/',
  source_2_Lazy = 'https://docs.sourceacademy.org/source_2_lazy/',
  source_3 = 'https://docs.sourceacademy.org/source_3/',
  source_3_Concurrent = 'https://docs.sourceacademy.org/source_3_concurrent/',
  source_3_Nondet = 'https://docs.sourceacademy.org/source_3_non-det/',
  source_4 = 'https://docs.sourceacademy.org/source_4/',
  source_4_Gpu = 'https://docs.sourceacademy.org/source_4_gpu/',
  ecmaScript_2021 = 'https://262.ecma-international.org/12.0/'
}

const Constants = {
  sourceAcademyVersion,
  sourceAcademyEnvironment,
  sourceAcademyDeploymentName,
  showResearchPrompt,
  backendUrl,
  cadetLoggerUrl,
  useBackend,
  defaultSourceChapter,
  defaultSourceVariant,
  defaultQuestionId,
  maxBrowseIndex,
  mobileBreakpoint,
  urlShortenerBase,
  urlShortenerSignature,
  moduleBackendUrl,
  authProviders,
  playgroundOnly,
  enableGitHubAssessments,
  sentryDsn,
  googleClientId,
  googleApiKey,
  googleAppId,
  githubClientId,
  githubOAuthProxyUrl,
  sharedbBackendUrl,
  disablePeriods,
  cadetLoggerInterval,
  sicpBackendUrl,
  workspaceSettingsLocalStorageKey,
  caFulfillmentLevel
};

export default Constants;
