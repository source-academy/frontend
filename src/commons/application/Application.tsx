import { Button, Classes, Dialog, H4, Intent } from '@blueprintjs/core';
import moment from 'moment';
import * as React from 'react';
import { Redirect, Route, RouteComponentProps, Switch } from 'react-router';

import Academy from '../../pages/academy/Academy';
import Contributors from '../../pages/contributors/Contributors';
import Disabled from '../../pages/disabled/Disabled';
import GitHubClassroom from '../../pages/githubAssessments/GitHubClassroom';
import GitHubCallback from '../../pages/githubCallback/GitHubCallback';
import Login from '../../pages/login/Login';
import MissionControlContainer from '../../pages/missionControl/MissionControlContainer';
import NotFound from '../../pages/notFound/NotFound';
import Playground from '../../pages/playground/PlaygroundContainer';
import Sicp from '../../pages/sicp/Sicp';
import Welcome from '../../pages/welcome/Welcome';
import { AssessmentConfiguration } from '../assessment/AssessmentTypes';
import NavigationBar from '../navigationBar/NavigationBar';
import Constants from '../utils/Constants';
import { useLocalStorageState } from '../utils/Hooks';
import { defaultWorkspaceSettings, WorkspaceSettingsContext } from '../WorkspaceSettingsContext';
import { Role } from './ApplicationTypes';
import { UpdateCourseConfiguration, UserCourse } from './types/SessionTypes';

export type ApplicationProps = DispatchProps & StateProps & RouteComponentProps<{}>;

export type DispatchProps = {
  handleLogOut: () => void;
  handleGitHubLogIn: () => void;
  handleGitHubLogOut: () => void;
  fetchUserAndCourse: () => void;
  handleCreateCourse: (courseConfig: UpdateCourseConfiguration) => void;
  updateCourseResearchAgreement: (agreedToResearch: boolean) => void;
};

export type StateProps = {
  role?: Role;
  name?: string;
  courses: UserCourse[];
  courseId?: number;
  courseShortName?: string;
  enableAchievements?: boolean;
  enableSourcecast?: boolean;
  assessmentConfigurations?: AssessmentConfiguration[];
  agreedToResearch?: boolean | null;
};

const loginPath = <Route path="/login" component={Login} key="login" />;

const Application: React.FC<ApplicationProps> = props => {
  const intervalId = React.useRef<number | undefined>(undefined);
  const [isDisabled, setIsDisabled] = React.useState(computeDisabledState());
  const isMobile = /iPhone|iPad|Android/.test(navigator.userAgent);
  const isPWA = window.matchMedia('(display-mode: standalone)').matches; // Checks if user is accessing from the PWA
  const browserDimensions = React.useRef({ height: 0, width: 0 });

  const [workspaceSettings, setWorkspaceSettings] = useLocalStorageState(
    Constants.workspaceSettingsLocalStorageKey,
    defaultWorkspaceSettings
  );

  const isLoggedIn = typeof props.name === 'string';
  const isCourseLoaded = isLoggedIn && typeof props.role === 'string';

  // Effect to fetch the latest user info and course configurations from the backend on refresh,
  // if the user was previously logged in
  React.useEffect(() => {
    if (props.name) {
      props.fetchUserAndCourse();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  React.useEffect(() => {
    if (Constants.disablePeriods.length > 0) {
      intervalId.current = window.setInterval(() => {
        const disabled = computeDisabledState();
        if (isDisabled !== disabled) {
          setIsDisabled(disabled);
        }
      }, 5000);
    }

    return () => {
      if (intervalId.current) {
        window.clearInterval(intervalId.current);
      }
    };
  }, [isDisabled]);

  /**
   * The following effect prevents the mobile browser interface from hiding on scroll by setting the
   * application height to the window's innerHeight, even after orientation changes. This ensures that
   * the app UI does not break due to the hiding of the browser interface when the user is not on the PWA.
   *
   * Note: When the soft keyboard is up on Android devices, the viewport height decreases and triggers
   * the 'resize' event. The conditional in orientationChangeHandler checks specifically for this, and
   * does not update the application height when the Android keyboard triggers the resize event. IOS
   * devices are not affected.
   */
  React.useEffect(() => {
    const orientationChangeHandler = () => {
      if (
        !(
          window.innerHeight < browserDimensions.current.height &&
          window.innerWidth === browserDimensions.current.width
        )
      ) {
        // If it is not an Android soft keyboard triggering the resize event, update the application height.
        document.documentElement.style.setProperty(
          '--application-height',
          window.innerHeight + 'px'
        );
      }
      browserDimensions.current = { height: window.innerHeight, width: window.innerWidth };
    };

    if (!isPWA && isMobile) {
      orientationChangeHandler();
      window.addEventListener('resize', orientationChangeHandler);
    }

    return () => {
      if (!isPWA && isMobile) {
        window.removeEventListener('resize', orientationChangeHandler);
      }
    };
  }, [isPWA, isMobile]);

  // Paths common to both deployments
  const commonPaths = [
    <Route path="/contributors" component={Contributors} key="contributors" />,
    <Route path="/callback/github" component={GitHubCallback} key="githubCallback" />,
    <Redirect
      from="/interactive-sicp/:section?"
      to="/sicpjs/:section?"
      key="oldToNewSicpRedirect"
    />,
    <Route path="/sicpjs/:section?" component={Sicp} key="sicp" />,
    Constants.enableGitHubAssessments ? (
      <Route
        path="/githubassessments"
        render={() => (
          <GitHubClassroom
            handleGitHubLogIn={props.handleGitHubLogIn}
            handleGitHubLogOut={props.handleGitHubLogOut}
          />
        )}
        key="githubAssessments"
      />
    ) : null
  ];

  const isDisabledEffective = !['staff', 'admin'].includes(props.role!) && isDisabled;

  return (
    <WorkspaceSettingsContext.Provider value={[workspaceSettings, setWorkspaceSettings]}>
      <div className="Application">
        <NavigationBar
          handleLogOut={props.handleLogOut}
          handleGitHubLogIn={props.handleGitHubLogIn}
          handleGitHubLogOut={props.handleGitHubLogOut}
          handleCreateCourse={props.handleCreateCourse}
          role={props.role}
          name={props.name}
          courses={props.courses}
          courseId={props.courseId}
          courseShortName={props.courseShortName}
          enableAchievements={props.enableAchievements}
          enableSourcecast={props.enableSourcecast}
          assessmentTypes={React.useMemo(
            () => props.assessmentConfigurations?.map(c => c.type),
            [props.assessmentConfigurations]
          )}
        />
        <div className="Application__main">
          {isDisabledEffective && (
            <Switch>
              {!Constants.playgroundOnly && loginPath}
              {/* if not logged in, and we're not a playground-only deploy, then redirect to login (for staff) */}
              {!isCourseLoaded && !Constants.playgroundOnly
                ? [
                    <Route path="/courses" render={redirectToLogin} key={0} />,
                    <Route exact={true} path="/" render={redirectToLogin} key={1} />
                  ]
                : []}
              <Route>
                <Disabled reason={typeof isDisabled === 'string' ? isDisabled : undefined} />
              </Route>
            </Switch>
          )}
          {!isDisabledEffective && Constants.playgroundOnly && (
            <Switch>
              {commonPaths}
              <Route path="/playground" component={Playground} />
              <Route exact={true} path="/">
                <Redirect to="/playground" />
              </Route>
              <Route component={NotFound} />
            </Switch>
          )}
          {!isDisabledEffective && !Constants.playgroundOnly && (
            <Switch>
              {loginPath}
              {commonPaths}
              <Route path={'/courses/:courseId(\\d+)?'} render={toAcademy(props)} />
              <Route path="/welcome" render={ensureUserAndRouteTo(props, <Welcome />)} />
              <Route
                path={'/mission-control/:assessmentId(-?\\d+)?/:questionId(\\d+)?'}
                component={MissionControlContainer}
              />
              <Route
                path="/playground"
                render={ensureUserAndRoleAndRouteTo(props, <Playground />)}
              />

              <Redirect
                from="/"
                exact={true}
                to={props.courseId != null ? `/courses/${props.courseId}` : '/welcome'}
              />
              {props.courseId != null && [
                <Redirect
                  from="/sourcecast/:splat?"
                  to={`/courses/${props.courseId}/sourcecast/:splat?`}
                  key="legacy-sourcecast"
                />,
                <Redirect
                  from="/achievements/:splat?"
                  to={`/courses/${props.courseId}/achievements/:splat?`}
                  key="legacy-achievements"
                />,
                <Redirect
                  from="/academy/:splat?"
                  to={`/courses/${props.courseId}/:splat?`}
                  key="legacy-academy"
                />
              ]}
              <Route component={NotFound} />
            </Switch>
          )}
        </div>

        {/* agreedToResearch has a default value of undefined in the store.
            It will take on null/true/false when the backend returns. */}
        {Constants.showResearchPrompt && props.agreedToResearch === null && (
          <div className="research-prompt">
            <Dialog
              className={Classes.DARK}
              title="Agreement to Participate in Educational Research"
              canOutsideClickClose={false}
              canEscapeKeyClose={false}
              isCloseButtonShown={false}
              isOpen
            >
              <div className={Classes.DIALOG_BODY}>
                <H4>Welcome to your new Source Academy @ NUS course!</H4>
                <div>
                  Here at Source Academy @ NUS, our mission is to bring out the beauty and fun in
                  programming and the ideas behind programming, and to make these ideas universally
                  accessible. This includes educational research!
                </div>
                <br />
                <div>
                  We collect programs that students run in Source Academy @ NUS and store them
                  anonymously for our research. You are free to opt out of this collection, with no
                  penalty for you whatsoever. Contact your course instructor if you have questions
                  or concerns about this research.
                </div>
              </div>
              <div className={Classes.DIALOG_FOOTER}>
                <div className={Classes.DIALOG_FOOTER_ACTIONS}>
                  <Button
                    text="I would like to opt out"
                    onClick={() => props.updateCourseResearchAgreement(false)}
                  />
                  <Button
                    text="I consent!"
                    intent={Intent.SUCCESS}
                    onClick={() => props.updateCourseResearchAgreement(true)}
                  />
                </div>
              </div>
            </Dialog>
          </div>
        )}
      </div>
    </WorkspaceSettingsContext.Provider>
  );
};

const redirectToLogin = () => <Redirect to="/login" />;
const redirectToWelcome = () => <Redirect to="/welcome" />;

/**
 * A user routes to /academy,
 *  1. If the user is logged in, render the Academy component
 *  2. If the user is not logged in, redirect to /login
 */
const toAcademy = ({ name, role }: ApplicationProps) =>
  name === undefined ? redirectToLogin : role === undefined ? redirectToWelcome : () => <Academy />;

/**
 * Routes a user to the specified route,
 *  1. If the user is logged in, render the specified component
 *  2. If the user is not logged in, redirect to /login
 */
const ensureUserAndRouteTo = ({ name }: ApplicationProps, to: JSX.Element) =>
  name === undefined ? redirectToLogin : () => to;

/**
 * Routes a user to the specified route,
 *  1. If the user is logged in and has a latest viewed course, render the
 *     specified component
 *  2. If the user is not logged in, redirect to /login
 *  3. If the user is logged in, but does not have a course, redirect to /welcome
 */
const ensureUserAndRoleAndRouteTo = ({ name, role }: ApplicationProps, to: JSX.Element) =>
  name === undefined ? redirectToLogin : role === undefined ? redirectToWelcome : () => to;

function computeDisabledState() {
  const now = moment();
  for (const { start, end, reason } of Constants.disablePeriods) {
    if (start.isBefore(now) && end.isAfter(now)) {
      return reason || true;
    }
  }
  return false;
}

export default Application;
