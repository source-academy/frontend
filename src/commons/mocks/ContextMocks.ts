import { parse } from 'acorn';
import { FunctionExpression, Node } from 'estree';
import { ACORN_PARSE_OPTIONS } from 'js-slang/dist/constants';
import createContext, { EnvTree } from 'js-slang/dist/createContext';
import Closure from 'js-slang/dist/interpreter/closure';
import { Context, Environment } from 'js-slang/dist/types';
import { TypeError } from 'js-slang/dist/utils/rttc';

export function mockContext(chapter = 1): Context {
  return createContext(chapter);
}

export function mockRuntimeContext(): Context {
  const context = createContext();
  // Note: noticed no harm in removing the following context.runtime.
  // If you get an error with head is undefined after trying to evaluate code.
  // Likely due to the environments: [].
  // In every real context, there is at least one env frame.
  context.runtime = {
    break: false,
    debuggerOn: true,
    isRunning: true,
    environmentTree: new EnvTree(),
    environments: [],
    nodes: [
      {
        type: 'Literal',
        loc: {
          start: { line: 1, column: 0 },
          end: { line: 1, column: 1 }
        },
        value: 0,
        raw: '0',
        range: [0, 1]
      }
    ],
    agenda: null,
    stash: null,
    envStepsTotal: 0,
    breakpointSteps: []
  };
  return context;
}

export function mockClosure(): Closure {
  return new Closure({} as FunctionExpression, {} as Environment, {} as Context);
}

export function mockTypeError(): TypeError {
  // Typecast to Node to fix estree-acorn compatability.
  return new TypeError(parse('', ACORN_PARSE_OPTIONS) as Node, '', '', '');
}
