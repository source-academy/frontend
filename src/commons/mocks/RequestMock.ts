import * as RequestsSaga from '../utils/RequestHelper';

export class RequestMock {
  static noResponse(): typeof RequestsSaga.request {
    return () => Promise.resolve(null);
  }

  static nonOk(textMockFn: jest.Mock = jest.fn()): typeof RequestsSaga.request {
    const resp = {
      text: textMockFn,
      ok: false
    } as unknown as Response;

    return () => Promise.resolve(resp);
  }

  static success(
    jsonMockFn: jest.Mock = jest.fn(),
    textMockFn: jest.Mock = jest.fn()
  ): typeof RequestsSaga.request {
    const resp = {
      json: jsonMockFn,
      text: textMockFn,
      ok: true
    } as unknown as Response;

    return () => Promise.resolve(resp);
  }
}

export const mockTokens = { accessToken: 'access', refreshToken: 'refresherOrb' };
