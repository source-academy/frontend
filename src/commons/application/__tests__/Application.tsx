import { shallow } from 'enzyme';
import moment from 'moment';
import Constants from 'src/commons/utils/Constants';

import { mockRouterProps } from '../../mocks/ComponentMocks';
import Application, { ApplicationProps } from '../Application';

// JSDOM does not implement window.matchMedia, so we have to mock it.
window.matchMedia =
  window.matchMedia ||
  function () {
    return {
      matches: false,
      addListener: function () {},
      removeListener: function () {}
    };
  };

const props: ApplicationProps = {
  ...mockRouterProps('/academy', {}),
  courses: [],
  handleLogOut: () => {},
  handleGitHubLogIn: () => {},
  handleGitHubLogOut: () => {},
  fetchUserAndCourse: () => {},
  updateLatestViewedCourse: () => {},
  handleCreateCourse: () => {},
  updateCourseResearchAgreement: () => {}
};

test('Application renders correctly', () => {
  const app = <Application {...props} />;
  const tree = shallow(app);
  expect(tree.debug()).toMatchSnapshot();
});

test('Application shows disabled when in disabled period', () => {
  const origPeriods = Constants.disablePeriods;
  const now = moment();
  Constants.disablePeriods = [
    {
      start: now.clone().subtract(1, 'minute'),
      end: now.clone().add(1, 'minute'),
      reason: 'Testing'
    }
  ];

  const app = <Application {...props} />;
  const tree = shallow(app);
  expect(tree.debug()).toMatchSnapshot();

  Constants.disablePeriods = origPeriods;
});
