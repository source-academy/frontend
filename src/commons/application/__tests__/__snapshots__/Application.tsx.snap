// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application renders correctly 1`] = `
"<ContextProvider value={{...}}>
  <div className=\\"Application\\">
    <NavigationBar handleLogOut={[Function: handleLogOut]} handleGitHubLogIn={[Function: handleGitHubLogIn]} handleGitHubLogOut={[Function: handleGitHubLogOut]} handleCreateCourse={[Function: handleCreateCourse]} role={[undefined]} name={[undefined]} courses={{...}} courseId={[undefined]} courseShortName={[undefined]} enableAchievements={[undefined]} enableSourcecast={[undefined]} assessmentTypes={[undefined]} />
    <div className=\\"Application__main\\">
      <Switch>
        <Route path=\\"/login\\" component={[Function: Login]} />
        <Route path=\\"/contributors\\" component={[Function: Contributors]} />
        <Route path=\\"/callback/github\\" component={[Function: GitHubCallback]} />
        <Redirect from=\\"/interactive-sicp/:section?\\" to=\\"/sicpjs/:section?\\" />
        <Route path=\\"/sicpjs/:section?\\" component={[Function: Sicp]} />
        <Route path=\\"/githubassessments\\" render={[Function: render]} />
        <Route path=\\"/courses/:courseId(\\\\\\\\d+)?\\" render={[Function: redirectToLogin]} />
        <Route path=\\"/welcome\\" render={[Function: redirectToLogin]} />
        <Route path=\\"/mission-control/:assessmentId(-?\\\\\\\\d+)?/:questionId(\\\\\\\\d+)?\\" component={[Function: C] { displayName: 'withRouter(Connect(MissionControl))', WrappedComponent: [Function: MissionControl], propTypes: { wrappedComponentRef: [Function: bound checkType] { isRequired: [Function: bound checkType] } } }} />
        <Route path=\\"/playground\\" render={[Function: redirectToLogin]} />
        <Redirect from=\\"/\\" exact={true} to=\\"/welcome\\" />
        <Route component={[Function: NotFound]} />
      </Switch>
    </div>
  </div>
</ContextProvider>"
`;

exports[`Application shows disabled when in disabled period 1`] = `
"<ContextProvider value={{...}}>
  <div className=\\"Application\\">
    <NavigationBar handleLogOut={[Function: handleLogOut]} handleGitHubLogIn={[Function: handleGitHubLogIn]} handleGitHubLogOut={[Function: handleGitHubLogOut]} handleCreateCourse={[Function: handleCreateCourse]} role={[undefined]} name={[undefined]} courses={{...}} courseId={[undefined]} courseShortName={[undefined]} enableAchievements={[undefined]} enableSourcecast={[undefined]} assessmentTypes={[undefined]} />
    <div className=\\"Application__main\\">
      <Switch>
        <Route path=\\"/login\\" component={[Function: Login]} />
        <Route path=\\"/courses\\" render={[Function: redirectToLogin]} />
        <Route exact={true} path=\\"/\\" render={[Function: redirectToLogin]} />
        <Route>
          <Disabled reason=\\"Testing\\" />
        </Route>
      </Switch>
    </div>
  </div>
</ContextProvider>"
`;
